File: src/cli/args.ts
import { readFileSync } from "node:fs";

export type CliOptions = {
  json: boolean;
  currency: "USD";
  dynadotKey: string | undefined;
  timeoutMs: number;
  domains: string[];
};

const usage = `Usage:
  domain search <domains...>
  domain search --file domains.txt

Flags:
  --json
  --currency USD
  --dynadot-key <key>
  --timeout 10s`;

const parseTimeout = (raw: string): number => {
  const seconds = raw.match(/^([0-9]+)s$/i);
  if (seconds?.[1]) return Number(seconds[1]) * 1000;

  const millis = raw.match(/^([0-9]+)ms$/i);
  if (millis?.[1]) return Number(millis[1]);

  const direct = Number(raw);
  if (Number.isFinite(direct) && direct > 0) return direct;

  throw new Error(`Invalid timeout value: ${raw}`);
};

const fromFile = (filePath: string): string[] =>
  readFileSync(filePath, "utf-8")
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0);

export const parseArgs = (argv: string[]): CliOptions => {
  const [command, ...rest] = argv;
  if (command !== "search") {
    throw new Error(usage);
  }

  const options: Omit<CliOptions, "domains"> = {
    json: false,
    currency: "USD",
    dynadotKey: undefined,
    timeoutMs: 10_000,
  };

  const positional: string[] = [];
  let filePath: string | null = null;

  for (let index = 0; index < rest.length; index += 1) {
    const token = rest[index];
    if (!token) continue;

    if (token === "--json") {
      options.json = true;
      continue;
    }

    if (token === "--file") {
      filePath = rest[index + 1] ?? null;
      index += 1;
      continue;
    }

    if (token === "--currency") {
      const value = rest[index + 1];
      if (!value || value !== "USD") throw new Error("Only USD is supported in v1");
      options.currency = value;
      index += 1;
      continue;
    }

    if (token === "--dynadot-key") {
      const value = rest[index + 1];
      if (!value) throw new Error("--dynadot-key requires a value");
      options.dynadotKey = value;
      index += 1;
      continue;
    }

    if (token === "--timeout") {
      const value = rest[index + 1];
      if (!value) throw new Error("--timeout requires a value");
      options.timeoutMs = parseTimeout(value);
      index += 1;
      continue;
    }

    positional.push(token);
  }

  const fromPath = filePath ? fromFile(filePath) : [];
  const domains = [
    ...new Set([...positional, ...fromPath].map((value) => value.trim().toLowerCase())),
  ].filter((value) => value.length > 0);

  if (domains.length === 0) throw new Error("No domains provided");

  return { ...options, domains };
};

File: internal/cli/args.go
// Package cli implements the dotld command-line interface.
package cli

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

type cliOptions struct {
	JSON       bool
	Currency   string
	DynadotKey string
	TimeoutMs  int
	Domains    []string
}

const usage = `Usage:
  dotld <domains...>
  dotld --file domains.txt

Flags:
  --json
  --currency USD
  --dynadot-key <key>
  --timeout 10s
  --version, -v`

var (
	secondsRe = regexp.MustCompile(`^([0-9]+)s$`)
	millisRe  = regexp.MustCompile(`^([0-9]+)ms$`)
)

func parseTimeout(raw string) (int, error) {
	if m := secondsRe.FindStringSubmatch(raw); len(m) > 1 {
		var n int
		fmt.Sscanf(m[1], "%d", &n)

		return n * 1000, nil
	}

	if m := millisRe.FindStringSubmatch(raw); len(m) > 1 {
		var n int
		fmt.Sscanf(m[1], "%d", &n)

		return n, nil
	}

	var n int
	_, err := fmt.Sscanf(raw, "%d", &n)
	if err == nil && n > 0 {
		return n, nil
	}

	return 0, fmt.Errorf("Invalid timeout value: %s", raw)
}

func fromFile(filePath string) ([]string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	var domains []string
	for _, line := range strings.Split(string(data), "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed != "" {
			domains = append(domains, trimmed)
		}
	}

	return domains, nil
}

var errVersion = fmt.Errorf("__version__")

func parseArgs(argv []string) (cliOptions, error) {
	if len(argv) > 0 && (argv[0] == "--version" || argv[0] == "-v") {
		return cliOptions{}, errVersion
	}

	if len(argv) == 0 || argv[0] != "search" {
		return cliOptions{}, fmt.Errorf("%s", usage)
	}

	rest := argv[1:]
	options := cliOptions{
		Currency:  "USD",
		TimeoutMs: 10_000,
	}

	var positional []string
	var filePath string

	for i := 0; i < len(rest); i++ {
		token := rest[i]
		if token == "" {
			continue
		}

		switch token {
		case "--json":
			options.JSON = true
		case "--file":
			if i+1 >= len(rest) {
				return cliOptions{}, fmt.Errorf("--file requires a value")
			}
			i++
			filePath = rest[i]
		case "--currency":
			if i+1 >= len(rest) {
				return cliOptions{}, fmt.Errorf("--currency requires a value")
			}
			i++
			if rest[i] != "USD" {
				return cliOptions{}, fmt.Errorf("Only USD is supported in v1")
			}
			options.Currency = rest[i]
		case "--dynadot-key":
			if i+1 >= len(rest) {
				return cliOptions{}, fmt.Errorf("--dynadot-key requires a value")
			}
			i++
			options.DynadotKey = rest[i]
		case "--timeout":
			if i+1 >= len(rest) {
				return cliOptions{}, fmt.Errorf("--timeout requires a value")
			}
			i++
			t, err := parseTimeout(rest[i])
			if err != nil {
				return cliOptions{}, err
			}
			options.TimeoutMs = t
		default:
			positional = append(positional, token)
		}
	}

	var fromPath []string
	if filePath != "" {
		var err error
		fromPath, err = fromFile(filePath)
		if err != nil {
			return cliOptions{}, err
		}
	}

	all := append(positional, fromPath...)
	seen := make(map[string]bool)
	var domains []string
	for _, v := range all {
		d := strings.TrimSpace(strings.ToLower(v))
		if d != "" && !seen[d] {
			seen[d] = true
			domains = append(domains, d)
		}
	}

	if len(domains) == 0 {
		return cliOptions{}, fmt.Errorf("No domains provided")
	}

	options.Domains = domains

	return options, nil
}
